ssl/quic/quic_rx_frame.c

```diff
@@ -0,0 +1,391 @@
+            stream_type = SSL_get_stream_type(stream);
+
+            /*
+             * ASSUMPTION: ssl_close_stream() detaches the stream SSL from
+             * the connection SSL it's attached to, and then destroys that
+             * stream SSL.
+             */
+            ssl_close_stream(stream);
+
+            if (stream_type == SSL_STREAM_TYPE_READ) {
+                /*
+                 * ASSUMPTION: ssl_close_connection() closes all the streams
+                 * that are attached to it, then destroys the connection SSL.
+                 */
+                ssl_close_connection(packet->connection);
+                packet->connection = NULL;
+                ossl_quic_set_state(???, QUIC_STREAM_STATE_ERROR);
+
+                /*
+                 * There's no point to continue with this packet.  However,
+                 * processing it was a success, so we return with a happy
+                 * face, and leave it the state machine what to do with the
+                 * general STREAM_STATE_ERROR state.
+                 */
+                return 1;
+            }
+            break;
+        }
+
+        case OSSL_QUIC_FRAME_TYPE_CRYPTO: {
+            OSSL_QUIC_FRAME_CRYPTO frame_data;
+
+            if (!ossl_quic_wire_decode_frame_crypto(pkt, &frame_data))
+                return 0;
+            /* ADD CODE to send |frame_data.data| to the handshake manager */
+            break;
+        }
+        case OSSL_QUIC_FRAME_TYPE_NEW_TOKEN: {
+            const uint8_t *token;
+            size_t token_len;
+
+            if (!ossl_quic_wire_decode_frame_new_token(pkt, &token, &token_len))
+                return 0;
+            /* ADD CODE to send |token| to the session manager */
+            break;
+        }
+
+        case OSSL_QUIC_FRAME_TYPE_STREAM:
+        case OSSL_QUIC_FRAME_TYPE_STREAM_FIN:
+        case OSSL_QUIC_FRAME_TYPE_STREAM_LEN:
+        case OSSL_QUIC_FRAME_TYPE_STREAM_LEN_FIN:
+        case OSSL_QUIC_FRAME_TYPE_STREAM_OFF:
+        case OSSL_QUIC_FRAME_TYPE_STREAM_OFF_FIN:
+        case OSSL_QUIC_FRAME_TYPE_STREAM_OFF_LEN:
+        case OSSL_QUIC_FRAME_TYPE_STREAM_OFF_LEN_FIN: {
+            OSSL_QUIC_FRAME_STREAM frame_data;
+
+            if (!ossl_quic_wire_decode_frame_stream(pkt, &frame_data))
+                return 0;
+
+            stream = ssl_get_stream(connection, frame_data.stream_id);
+
+            /*
+             * ASSUMPTION: ssl_queue_data() adds data to an stream SSL,
+             * to be consumed by the application when doing an SSL_read()
+             *
+             * It's currently not determined what that queueing does, if
+             * it copies the data into another buffer, or if that should
+             * attach the OSSL_QRL_RX_PKT to the stream SSL queue, and
+             * exactly what that entails in terms of deallocation of that
+             * packet.
+             */
+            if (!ssl_queue_data(stream, frame_data.data, frame_data.len))
```

Comment:
I expect the stream data to not be copied into another buffer, but be kept around until it is no longer needed. This will mean we can't release the packet until all streams have finished with the data. We might need to do some ref counting on some quic packet object.

Reply:
Agreed...  but how remains to be determined.

Reply:
We might want to have a sort of buffer metadata structure with whatever we want (such as a refcount), so we can pass it around without having to worry too much all over the place.  The RX depacketizer code just passes the data on, it shouldn't have to worry about refcounts, but since `ssl_queue_data()` (or whatever that function will end up being called) would grab a reference to that buffer, it might be interested in incrementing the refcount.

Reply:
> The RX depacketizer code just passes the data on, it shouldn't have to worry about refcounts, but since ssl_queue_data() (or whatever that function will end up being called) would grab a reference to that buffer, it might be interested in incrementing the refcount.

The RX depacketizer code needs to be aware of this to some extent - since it is that code that calls `ossl_qrx_release_pkt` (at the moment at least). That function call will release the underlying buffer, so it must not be called until everything that is using it has finished with it.

Reply:
I was thinking that the the OSSL_QRX_PKT could hold a pointer to that metadata structure instead of directly to the actual data, and that the metadata structure would have pointers to the actual data.  `ossl_qrx_release_pkt()` would then be able to decrement the refcount and all other magic as it likes.

BTW, the `OSSL_QRX_PKT` has a magic opaque pointer...  could it be that we already have the necessary stuff to do that magic?  @hlandau ?

Reply:
Not sure why we wouldn't just ref count the `OSSL_QRX_PKT` structure.

Reply:
> And I'm not sure why ossl_qrx_release_pkt() can't do that, then...

Because I anticipate the record layer being something that we will eventually make pluggable/fetchable. I think we should keep the interface to the record layer as simple as possible.

Reply:
> > And I'm not sure why ossl_qrx_release_pkt() can't do that, then...
> 
> Because I anticipate the record layer being something that we will eventually make pluggable/fetchable. I think we should keep the interface to the record layer as simple as possible.

Do you meant to say that this code would call the plugged funtion directly (i.e. something along the lines of `qrx->release_pkt()`, or would `ossl_qrx_release_pkt()` still be used as a wrapper?  In the latter case, I fail to see the problem.

Reply:
The QRX decrypts entire packets. Once you call `ossl_qrx_release_pkt`, you may no longer reference the decrypted buffer it provides you. (The QRX would be entitled to then free the buffer, in practice it will recycle it for a new packet.)

If there is data in the packet you need to keep for an extended period of time, you need to either copy it (which we are trying to avoid), or avoid releasing the packet until you no longer need to refer to *any* data within it. There is no time limit on this; if all of the QRX's internal buffers are checked out and you try to receive new packets, it will just allocate more.

My assumption here is that there is bookkeeping the RX depacketiser will want to do which means that it will want to define its own internal bookkeeping structure. There is no need to keep the `OSSL_QRX_PKT` structure itself around so long as you remember the `handle` field so you can eventually pass it to `ossl_qrx_release_pkt`. It would probably be logical for the RX depacketiser to keep a refcount to determine when to call `ossl_qrx_release_pkt` if it has complex lifetime requirements.

Reply:
Ok...  I was imagining that the RX depacketizer wouldn't have to do that...  Meh...

Reply:
I've added bookkeeping code, as a wrapper around the `OSSL_QRX_PKT` handle.